---
title: "Epigenomic Aging"
author: "Eric Kramer"
date: "1 June 2015"
output: rmarkdown::tufte_handout
---

```{r echo=F, include=FALSE}
library("dplyr")
library("tidyr")
library("knitr")
library("xtable")
library(magrittr)
library("pcaMethods")
library(ggplot2)
library("glmnet")
library(VennDiagram)
library("data.table")
library("gsl")

options(xtable.comment = FALSE)
options(xtable.booktabs = TRUE)

load("../data/sample.info.Rdata")
load('../data/aimms.Rdata')
load("../data/ancestry_model.Rdata")
load("../data/predicted.ancestry.Rdata")
load("../data/horvath_ages.Rdata")
load("../data/age.Rdata")
load("../data/tissue.Rdata")
load("../data/afr.Rdata")

acc = predicted.ancestry %>%
  filter(!is.na(ancestry)) %>%
  filter(ancestry %in% c("EUR", "AFR", "ASN")) %>%
  group_by(series.id) %>%
  summarize(acc=sum(ancestry == predicted.ancestry) / n(),
            N=n())

sig = function(x) x %>%
  filter(q.value < 0.05) %>%
  filter(dBeta > 0.05) %>% 
  as.data.frame %>%
  nrow

source("../figures/gene_plot.R")

```

# Introduction

DNA methylation is the binding of a methyl group to a guanine on DNA. Multiple studies have documented correlations between methylation in a gene's promoter and the expression of that gene; however, the exact causal relationship between these variables is unknown and likely depends on a variety of factors, such as the presence of eQTLs and the density of CpG sites in the region.

DNA methylation plays a crucial role in tissue differentiation. Studies have tracked increasing methylation as a cell proceeds down the differentiation pathway for innate immune cells, while the first principle component of methylation data will separate tissues, if there are multiple tissues present. 

More recent work has shown that environmental factors also affect DNA methylation. Age, socio-economic status and ancestral background can all affect DNA methylation. ^[Disentangling the effects of ancestry and socio-economic presents problems in many datasets where only one of these variables is measured although they are strongly correlated] In general, studies have shown increased methylation with age and low socio-economic status. Nseuronal pathways appear to be the most highly methylated with agining; while stress and immune pathways demonstrate the most increase in methylation due to low socio-economic status.

Hovarth's seminal 2013 paper provided a predictive model of age based on DNA methylation. This model is able to predict a sample's age with astonishing accuracy. More importantly, the model is applicable to a wide-range of tissues. However, we noticed that this model performs worse on samples of African ancestry, likely due to the above-mentioned environmental effects. Because of this, we decided to recalibrate the epigenetic so that is provides improved performance in non-Europeans, while still maintaining performance in diverse tissues.

The following work is an attempt to rebuild the epigenetic clock of Horvath to ensure that the model performs equally well on European, African and Asian samples; while still demonstrating the cross-tissue validity of the original epigenetic clock.

\begin{marginfigure}
\includegraphics{horvath_clock.png}
\caption{Horvath's epigenetic clock performs significantly worse on Belgians of african ancestry (AFB) than Belgians of European ancestry (EUB)}
\end{marginfigure}

This work proceeeds in four primary steps:

1. Preprocessing of publically available methylation data
2. Creation of predictive model for ethnicity
3. Promoter-level feature selection
4. Recalibrarting the epigenetic clock

# Preprocessing of publically available methylation data

In order to recalibrate the epigenetic clock so that is stable across ancestries and tissues, we require a training set with samples from multiple ancestries and multiple tissues.

## Gene Expression Omnibus data

Gene Expression Omnibus (GEO) contains over 20,000 Illumina 450k DNA methylation arrays. Among these, approximately `r nrow(sample.info)` samples are annotated with an age. I downloaded and processed raw data for these GEO series and placed these data into a SQLite database ^[See ./data/BMIQ.db. Recent work has allowed SQLite databases to be used within R as $data.frames$. See the $dplyr$ package for more detail]. These samples range in age from newborns to `r max(sample.info$age, na.rm=T)` years old.

```{r, fig.width = 10, fig.height = 4, fig.cap = "Age distribution of GEO samples assembled for this project", echo=F, warning=FALSE, fig.fullwidth = TRUE}

ggplot(sample.info,
       aes(age)) +
  geom_histogram(binwidth=2) +
  ylab("Count") +
  xlab("Age")
```


This database contains samples from `r length(unique(sample.info$tissue))` tissues. A majority of the samples are whole blood or other blood-based cell types (e.g. monocytes). While solid-tissue samples, such as samples from lung or muscle, are comparatively more rare. Because of the highly unbalanced nature of the dataset, careful statistics must be used to ensure that measured effects are not caused by confounding.

```{r, echo=FALSE, fig.margin=T}

tissue_counts = sample.info %>%
  group_by(tissue) %>%
  summarize(Count=n()) %>%
  select(Tissue=tissue, Count) %>%
  as.data.frame

kable(tissue_counts, caption="Number of samples for each tissue")

```



## Normalization 

BMIQ normalization is one of the most common methods to normalize DNA methylation data. It uses a mixture of beta-distributions and quantile normalization to adjust for differences between the probe types on the chip. After normalization, we see that the PCA nicely separates samples based on tissue. The first principle component separates liquid tissues (whole blood, monocytes, T-cells, leukocytes, etc) from the solid tissues (lung, brain, muscle, pancreas, etc). The second principle component is correlated with age and could represent age-related changes in methylation. 

```{r echo=FALSE, fig.cap="PCA of GEO samples colored according to tissue group", fig.fullwidth = TRUE, fig.width = 10, fig.height = 5,}

load("./pcas/scores.no.lymphoblasts.Rdata")

scores %<>% 
  mutate(`Tissue Group`=ifelse(tissue %in% c("Leukocytes", "Monocytes", "T-cells", "Whole Blood"),
                     "Liquid",
                     "Solid"))

ggplot(scores, aes(PC1, PC2, color=`Tissue Group`)) +
  geom_point()

```

If we color the plot according to tissue, rather that tissue group, we see that the two outlying clusters of solid tissues are samples from the cerebellum and the brain, excluding the cerebellum. Additionally, we see sub-structure within each of the clusters. For instance, monocytes in blue and T-cells in purple are clearly separated on the PCA.

```{r echo=FALSE, fig.cap="PCA of GEO samples colored according to tissue", fig.fullwidth = TRUE, fig.width = 10, fig.height = 5,}

load("./pcas/scores.no.lymphoblasts.Rdata")

scores %<>% 
  mutate(`Tissue Group`=ifelse(tissue %in% c("Leukocytes", "Monocytes", "T-cells", "Whole Blood"),
                     "Liquid",
                     "Solid"))

ggplot(scores, aes(PC1, PC2, color=tissue)) +
  geom_point()

```


# Creation of a predictive model for ancestry

Unfortunately, most GEO entires are not annotated with ancestral background. Because of this, we decided to create a predictive model, which is able to impute a sample's ancestry based on it's DNA methylation levels. To do this, we used methylation sites where common SNPs are located. These methylation sites allow us to infer genotype of a sample, and therefore the ancestry.

## Ancestry Informative Methylation Markers (AIMMs)

DNA methylation can only occur at guanines immediately preceeded by a cystine, known as CpG sites. SNPs in a CpG site will necessarily disrupt methylation. Conversely, we can use methylation at sites with known SNPs as a way to infer genotypes. 

In order to create a set of ancestry informative methyaltion markers, I mapped common variants from the 1000 genomes project (>5% MAF in one population) onto the CpG sites on the Illumina 450k array. In total, I discovered `r length(aimms)` methylation sites with a common SNP in the CpG site. ^[The ids for these sites are stored in ./data/aimms.Rdata in the project directory.]

## Training the ancestry model 

Data from the Human Variation Panel was used for training the model. This dataset contains 96 samples from each of three ancestral background: European (EUR), African (AFR) and Asian (ASN). All samples were collected in the United States. Importantly, the samples are lymphoblastic cell lines. Because of this, the samples appear to be over 100 years-old according to the epigenetic clock.

Elastic net regression was used to build a predictive model for ancestry ^[See ./ancestry_prediction/aimm_model.R for script for training the model]. The model selected `r sum(coef(m)[[1]] != 0) - 1` probe sites for inclusion in the model.

## Testing the ancestry model

Four GEO series, whose samples were annotated with ancestry information, were used to test the ancestry predictive model. The accuracy on these sets ranged from `r round(min(acc$acc), 2)` to `r max(acc$acc)`. We see that the ancestry model performs well on all four of these datasets.

```{r, echo=FALSE}

series.info = sample.info %>%
  group_by(series.id) %>%
  summarize(Tissue=tissue[1],
            `Average Age`=mean(age, na.rm=T))

acc %>%
  inner_join(series.info, by="series.id") %>%
  select(`GEO Series`=series.id, Tissue, `Average Age`, N, Accuracy=acc) %>%
  kable(caption="Accuracy of Ancestry Model", digits=2)

```

Hispanic samples represent an interesting test case for the ancestry model both because this population was not included in the training set and because the hispanic population is admixed. We see that hispanic samples are predicted to be either European or Asian.

```{r echo=F, fig.margin=T, fig.cap="Most likely ancestry for Hispanic samples"}

hsn = predicted.ancestry %>%
  filter(ancestry=="HSN") %>%
  mutate(`Predicted Ancestry`=predicted.ancestry)

ggplot(hsn, aes(`Predicted Ancestry`, fill=`Predicted Ancestry`)) +
  geom_bar() +
  ylab("Count")

```

Furthermore, if we look at the probabilities for each ancestry for each of the hispanic samples, we see that the model predicts some hispanic samples to be almost entirely Asian, while others are predicted to be almost entirely European. 

```{r, echo=F, fig.width=10, fig.height=4, fig.fullwidth = TRUE, fig.cap="Predicted Ancestral Probabilities of Hispanic Samples"}
hsn = predicted.ancestry %>%
  filter(ancestry=="HSN") %>%
  arrange(EUR) %>%
  mutate(id=1:nrow(.)) %>%
  select(id, AFR, ASN, EUR) %>%
  gather(ancestry, p, -id) %>%
  mutate(Ancestry=factor(ancestry))

suppressWarnings(print(
  ggplot(hsn, aes(id, p, fill=Ancestry)) +
  geom_bar(stat="identity") +
  theme(axis.title.x=element_blank()) +
  ylab("Probability of Ancestry") +
  xlab("Sample")
))

```

It would be interesting to correlated these predicted ancestry probabilities with actual admixture estimates. It is possible that these predicted probabilities correlated well with admixture estimates for the proportion of European and Asian (that is Native American) ancestry in these hispanic samples.

## Applied ancestry model to entire GEO dataset

I then applied the ancestry model to the entire GEO dataset. Unsurprisingly, we find that most of the samples in GEO are of European origin, followed by African and then Asian samples as the most rare.

```{r echo=F}

tmp = sample.info %>%
  inner_join(predicted.ancestry %>% select(-ancestry), by=c("gsm.id", "series.id")) %>%
  filter(tissue!="Lymphocytes") %>% # this series is an outlier, needs to be discarded
  group_by(tissue, predicted.ancestry) %>%
  summarize(n=n()) %>%
  spread(predicted.ancestry, n, fill=0) %>%
  mutate(Total=AFR + ASN + EUR)

col_counts = data.frame(tissue="Total",
                        AFR=sum(tmp$AFR),
                        ASN=sum(tmp$ASN),
                        EUR=sum(tmp$EUR),
                        Total=sum(tmp$Total))

rbind(tmp, col_counts) %>%
  select(Tissue=tissue, AFR, ASN, EUR, Total) %>%
  kable(caption="Counts for each tissue-ancestry pairing")

```

# Promoter-level analysis

Next, I used linear modeling to determine the effects of age, tissue and ancestry on methylation for each promoter of `r length(unique(age$gene))` genes. Importantly, these models look for broad effects. For example, measurement of the effect of aging is an effect with is stable across multiple probe sites, multiple tissues and multiple ethnicity. The creation of a dataset that is this large allows us to measure these effects in a highly accurate manner.


## Preprocessing probes

Before running the linear models, I discarded probe sites which have a common SNP in the sequence of the probe, multiple matching probes and probes on the X and Y chromosome. I defined the promoter region as the segment within 1,500 base pairs of the transcription start site (TSS). For each gene, if there were multiple transcription start sites, I took the TSS with the greatest number of methylation probes. Finally, I discarded genes with fewer than 4 sites in their promoter. In total, this left `r length(unique(age$gene))` genes ^[See ./linear_modeling/util.R for the script which conducts this probe filtering].

## Preprocessing samples

Before running the model, I discarded samples from $GSE56105$ as this GEO series showed a batch effect on principle component analysis, as well as lymphoblastic cell samples. Additionally, I applied Horvath's epigenetic clock to the dataset and discarded `r sum(horvath_ages$horvath_age > 120)` samples with an age greater than 120 as these are likely cancerous samples.

## Preprocessing tissue annotations

I decided to group the tissues into two main categories: solid and liquid tissues. These two main categories correspond to the two main clusters shown in the PCA in figure 2. In the linear model, I measure da fixed effect, which corresponds to the methylation differences between these two tissue groups, and I controled for the substructure within each of these groups using random effects.

## Fitting models

For each of these promoters, I ran a separate linear mixed effects model ^[See ./linear_modeling/stratified_model_funcs.R for the function which fits one of these linear mixed effects models] for each promoter ^[See ./linear_modeling/stratified_models.R for a wrapper to distribute this computation on BIC]. Each model calculated the effect of of age, tissue group and ancestry on methylation across all probes in the promoter. 

## Results

I defined significant differentially methylated promoters (DMPs) as those that showed both statistical significance (FDR p-value < $0.05$) and biological significant ($\Delta \beta > 0.05$). This resulted in 375 tissue DMPs, 353 age DMPs and 163 ancestry DMPs. 


```{r echo=F}
tissue %>% 
  filter(q.value < 0.05) %>%
  arrange(-abs(dBeta)) %>%
  select(Gene=gene, Estimate, `$\\Delta \\beta$`=dBeta, `q value`=q.value) %>%
  head(10) %>%
  kable(digits=2, caption="Top Tissue DMPs")

```

```{r echo=FALSE, fig.margin=TRUE, fig.width=5, fig.height=5, fig.cap="Venn Diagram of differentially methylated promoters (DMPs)"}

age = age %>% as.data.frame
tissue = tissue %>% as.data.frame
afr = afr %>% as.data.frame

filter = dplyr::filter
select = dplyr::select



sig.genes = function(x) x %>%
  filter(q.value < 0.05) %>%
  filter(dBeta > 0.05) %>%
  .$gene

venn.diagram(list("Age DMPs"=sig.genes(age),
                  "Tissue DMPs"=sig.genes(tissue),
                  "Ancestry DMPs"=sig.genes(afr)),
             filename=NULL,
             fill=c("blue","red", "green")) %>%
  grid.draw

```

```{r echo=F, fig.cap="Top Age DMPs"}
age %>% 
  filter(q.value < 0.05) %>%
  arrange(-abs(dBeta)) %>%
  select(Gene=gene, Estimate, `$\\Delta \\beta$`=dBeta, `q value`=q.value) %>%
  head(10) %>%
  kable(digits=2, caption="Top Age DMPs")

```



```{r echo=F, fig.cap="Top Ancestry DMPs"}
afr %>% 
  filter(q.value < 0.05) %>%
  arrange(-abs(dBeta)) %>%
  select(Gene=gene, Estimate, `$\\Delta \\beta$`=dBeta, `q value`=q.value) %>%
  head(10) %>%
  kable(digits=2, caption="Top Ancestry DMPs")

```



# Recalibrating the epigenetic clock

To recalibrate the epigenetic clock, I used the age-DMPs as defined by the previous section and use the probes in those promoters to build a predictive model for age based on methylation. I decided to use an ensemble approach for this modeling. I used 80% of the data to train 10 different predictive models using 10 different algorithms. Then, I used 10% of the data (not included in the initial training set) to build an ensemble from these models. In the ensemble, a final prediction is made by taking the weighted average of the predictions from each of the 10 models used as the basis.

```{r echo=FALSE}

cols = c(ppr="Regression",
         svmLinear="Kernel",
         pls="Regression",
         rvmLinear="Kernel",
         gbm="Boosting",
         rvmRadial="Kernel",
         pcr="Regression",
         svmRadialCost="Kernel",
         gaussprRadial="Kernel",
         rf="Boosting",
         ensemble="Ensemble",
         cubist="Boosting")

models = data.frame(Abbreviation=names(cols),
                    Name=c("Principle Pursuit Regression",
                           "Support Vector Machine - Linear Kernel",
                           "Partial Least Squares",
                           "Revelevance Vector Machine - Linear Kernel",
                           "Stochastic Gradient Boosting",
                           "Reveleance Vector Machine - Radial Kernel",
                           "Principle Component Regression",
                           "Support Vector Machine - Radial Kernel",
                           "Gaussian Process",
                           "Random Forest",
                           "Ensemble Model",
                           "Cubist Regression Tree"),
                    Family=cols) %>%
  filter(Family != "Ensemble")

kable(models %>% arrange(Family), caption="Predictive modeling methods used in ensemble model")

```

## Optimizing base models via cross-validation

Many of the base models contain parameters, which need to be optimized for good performance. Thus, for each of the models, I conducted a 10-fold cross-validation with a grid search over reasonable parameters. The set of parameters which showed the best performance for each model during this cross-validation was then used. ^[See ./model_building/get_params.R for a script which has reasonable parameter spaces for each of the models used in the ensemble]

## Training ensemble model



## Test final result

```{r echo=FALSE, fig.width=5, fig.height=5}

load("../data/ensemble.Rdata")

y3 = data3$age.normed

cols = c(ppr="Regression",
         svmLinear="Kernel",
         pls="Regression",
         rvmLinear="Kernel",
         gbm="Boosting",
         rvmRadial="Kernel",
         pcr="Regression",
         svmRadialCost="Kernel",
         gaussprRadial="Kernel",
         rf="Boosting",
         ensemble="Ensemble",
         cubist="Boosting")

perf = cor(cbind(z3, y3.pred), y3) %>%
  as.data.frame %>%
  mutate(method=row.names(.)) %>%
  mutate(cor=V1) %>%
  mutate(family=cols[method]) %>%
  arrange(family, cor) %>%
  mutate(method=factor(method, levels=method))

  
ggplot(perf, aes(method, V1, fill=family)) + 
  geom_bar(stat="identity") +
  theme_bw() +
  ggtitle("Performance on Testing Set")+
  ylab("Correlation between observed and predicted age") +
  xlab("Prediction Method") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
 


```

```{r echo=F, fig.width=5, fig.height=5}
  
age.antitrans = function(x) 21 * lambert_W0(exp(21*x + 1)^(1/21)) - 1

scatter.data = cbind(y3.pred, y3) %>%
  as.data.frame %>%
  mutate(age.pred=age.antitrans(ensemble)) %>%
  mutate(age=age.antitrans(y3))

ggplot(scatter.data, aes(age, age.pred)) +
  geom_point() +
  theme_bw() +
  xlab("Age") +
  ylab("Predicted Age") +
  ggtitle("Observed vs Predicted Age: Testing Set")
```

